### ISSUE TITLE
[Question] Ability to assign column types by column name (header) when importing CSV

### ISSUE THREAD

**Issue by zsoki** [jtablesaw/tablesaw#662]
Hi All,

First of all, really great work on Tablesaw, it's a joy to use so far!

I have a use case for which I yet to find a convenient solution: I have a CSV I need to import, which could have columns I don't need to process. However, I'd like to set the column types of the needed columns explicitly by 
```
val csvReadOptions = CsvReadOptions.builder(inputStream)
                .header(true)
                .columnTypes(columnTypeArray)
Table.read().csv(csvReadOptions)
```
The problems are the following:
- There could be columns missing
- The order of the columns could be different each time
- To increase performance, it would be ideal to skip all unknown columns not defined

This means the only thing I could determine the column types dynamically is if I could map the types to column names (headers).

My question is: is there a built-in functinality in Tablesaw for something like this?
```
val columnTypeMapping = mapOf(
    "column_name_1" to ColumnType.DOUBLE,
    "column_name_2" to ColumnType.STRING
)

val csvReadOptions = CsvReadOptions.builder(inputStream)
    .header(true)
    .columnTypeMapping(columnTypeMapping)
```

The only workaround I've could think of so far is the following:
- Read the first line of the CSV manually to get the headers, create the mapped ColumnType array manually, then reset the InputStream to feed into Tablesaw's CsvReader

I could also read the CSV twice with Tablesaw and get the headers from the Table but it seems a "little bit" wasteful 🙂

Also If you have any other ideas how to achieve this, I'd be glad to hear. Thanks!

**Comment by lwhite1** [comment]
@Suppoze Thanks for the kind words

This is kind of a tricky case because of the variability between file versions. 

I'm not sure I completely understand what you're trying to do, but i might attempt writing a method that:
- reads a "baseline" file. This has the columns you want 
- uses table.structure() to get a table containing the names and types.
- saves the structure table as a CSV file

Then write a second method that:
- reads the baseline structure table.
- get the headers [1] from your new file
- create a column type array to match the new file structure, using the SKIP columnType for any columns you want to skip, and using the data from your structure table to provide the other column types by matching on the headers
- load your file

[1] We have a method to just get the headers for a file, but it's private. It's in FileReader. If you want to submit a PR for making that method public that would be fine.

**Comment by lwhite1** [comment]
@zsoki LMK if you have any additional concerns. Otherwise I will close this in a couple days.

**Comment by zsoki** [comment]
@lwhite1 
Thank you for the detailed response!

TL;DR: Feel free to close as we went with another approach for the problem. Below I'll go into it in more detail 🙂

What you proposed is a great idea, thanks! However we ended up not needing this kind of functionality, at least for now. Instead we chose to go with an approach which simply handles the inconsistencies. For closure's sake, here's what we ended up with:

1. When reading the from the CSV, we simply use Tablesaw's automatic column type detection. Let's call it `inputTable`
2. We have a configuration which tells the program which columns are expected, without their type
3. We try to get each expected column from `inputTable` as generic `Column<*>` -> if it doesn't exist, we handle the exception, and the columns we don't need get discarded (this would be the SKIP part)
4. The configuration also has calculations defined which use the previous expected columns as input. The calculations do an instanceof check on the input columns (`Column<*>`) beforehand and we handle the exception if any is incorrect - so the type expectation is handled later by the calculators instead of when reading from the CSV

This was a sufficient solution because at this point we only need calculators for either `StringColumn` or `NumberColumn` input types - we don't need  (or rather, can't) define precision for the CSV columns with numerical data.

I do think we could improve upon this by using a predefined 'baseline' table, and the `headers` method would be a great utility in this and other use cases as well. I'll consider making a PR. 🙂

Right now I don't have anything to add to the topic so feel free to close.

Thanks again and sorry for the late reply!

**Comment by bantu** [comment]
I've got a related problem. I am reading three csv files into one table each. I want to append them all into one large table, but this is not possible because the types inferred by tablesaw are different. Manually specifying all types is not an option, since a lot of the data I do not need and the order of columns might change. Ideally, I would like to do the following for each table:
* Read csv into string table
* Retain only relevant data and change order of columns
* Change column types to my definition

**Comment by lwhite1** [comment]
Too late to matter here, but for posterity. The easiest way to force three tables to have the same structure is to use the columnTypes of the first to set the columnTypes for the others. 

This code may not be exactly right, but its in the right ballpark. 
```
   Table.read().csv(CsvReadOptions.builder().columnTypes(myOtherTable.columnTypeArray()))...
```
